### PR TITLE
Support parameterized paths

### PR DESCRIPTION
**Pull request by tatgean**
**Add support to paths like this:**
`/downloads/:platform/:version`

**In skygear js cloud code:**
`skygearCloud.handler('downloads/:platform/:version', ......)`

**Request:**
`/downloads/osx/1.0.0`

**In the handler function:**
`req.params.platform === 'osx'`
`req.params.version === '1.0.0'`

### CODE REVIEW

**Comment by oursky-travis** [comment]
@tatgean, thanks for your PR! By analyzing the history of the files in this pull request, we identified @rickmak to be a potential reviewer.

**Comment by tatgean** [comment]
Need this feature in order to work on [this issue](https://github.com/skygear-demo/skygear-squirrel-server/issues/1).
Thanks! :)

**Comment by chpapa** [comment]
Oh thanks for taking the initiative @tatgean :) 

Would be great if you can open a feature issue next time :) So we make sure we will get the same feature on python, and properly document it. I have put it up for you: SkygearIO/features#78

**Comment by tatgean** [comment]
Got it. Thank you!

File: packages/skygear-core/lib/cloud/registry.js [hunk 1]
```diff
@@ -143,7 +143,32 @@ export class Registry {
...
   getHandler(name, method) {
-    return this.handlers[name][method];
+    for (let k in this.handlers) {
+      if (this._matchHandler(k, name))
+        return this.handlers[k][method];
+    }
+    return undefined;
+  }
+
+  parseParamsInPath(path) {
```
Comment (rickmak) [inline packages/skygear-core/lib/cloud/registry.js]: The registry responsibility is not parsing the incoming request.

`parseParamsInPath` belong to `SkygearRequest` should make more semantic sense. Take a look at current implementation of `SkygearRequest` constructor, it already contains the logic of parsing incoming request.

File: packages/skygear-core/lib/cloud/registry.js [hunk 2]
```diff
@@ -143,7 +143,32 @@ export class Registry {
...
+          if (handlerParts[i].startsWith(':'))
+            params[handlerParts[i].replace(':', '')] = pathParts[i];
+        }
+        return params;
+      }
+    }
+    return undefined;
+  }
+
+  _matchHandler(handlerName, path) {
```
Comment (rickmak) [inline packages/skygear-core/lib/cloud/registry.js]: More test case on `_matchHandler` alone will make this PR better.

**Comment by rickmak** [comment]
hi @tatgean I researched some popular framework on both js and python. And propose to use the format `/downloads/{platform}/{version}`. Please take a look: https://github.com/SkygearIO/features/issues/78#issuecomment-309001643

And let me know do you agree on this? We can discuss and process with this new feature.

**Comment by chpapa** [comment]
@rickmak  this seems dragged a bit long, is it possible for you to actually fix and merge it?

**Comment by rickmak** [comment]
To be follow up at https://github.com/SkygearIO/skygear-SDK-JS/pull/344

**Pull request closed without merging**